rasa/cli/data.py

```diff
@@ -234,6 +273,422 @@ def split_nlu_data(args: argparse.Namespace) -> None:
+        args.nlu_classification_report
+    )
+    random.seed(args.random_seed)
+
+    # Determine low data, low performing and frequently confused intents
+    num_intents = len(nlu_training_data.intents)
+    avg_size = int(
+        reduce(
+            lambda acc, num: acc + (num / num_intents),
+            nlu_training_data.number_of_examples_per_intent.values(),
+            0,
+        )
+    )
+
+    low_data_intents = sorted(
+        nlu_training_data.number_of_examples_per_intent.items(),
+        key=operator.itemgetter(1),
+    )[: args.num_intents]
+    low_precision_intents = sorted(
+        map(
+            lambda k: (k, classification_report[k]["precision"]),
+            nlu_training_data.intents,
+        ),
+        key=operator.itemgetter(1),
+    )[: args.num_intents]
+    low_recall_intents = sorted(
+        map(
+            lambda k: (k, classification_report[k]["recall"]), nlu_training_data.intents
+        ),
+        key=operator.itemgetter(1),
+    )[: args.num_intents]
+    low_f1_intents = sorted(
+        map(
+            lambda k: (k, classification_report[k]["f1-score"]),
+            nlu_training_data.intents,
+        ),
+        key=operator.itemgetter(1),
+    )[: args.num_intents]
+    freq_confused_intents = sorted(
+        map(
+            lambda k: (k, sum(classification_report[k]["confused_with"].values())),
+            nlu_training_data.intents,
+        ),
+        key=operator.itemgetter(1),
+        reverse=True,
+    )[: args.num_intents]
+
+    pooled_intents = (
+        set(map(lambda tp: tp[0], low_data_intents))
+        | set(map(lambda tp: tp[0], low_precision_intents))
+        | set(map(lambda tp: tp[0], low_recall_intents))
+        | set(map(lambda tp: tp[0], low_f1_intents))
+        | set(map(lambda tp: tp[0], freq_confused_intents))
+    )
+
+    # Retrieve paraphrase pool and training data pool
+    paraphrase_pool = _create_paraphrase_pool(
+        paraphrases, pooled_intents, args.paraphrase_score_threshold
+    )
+    training_data_pool, training_data_vocab_per_intent = _create_training_data_pool(
+        nlu_training_data, pooled_intents
+    )
+
+    # Build augmentation pools based on the maximum vocabulary expansion criterion ("diverse") and random sampling
+    max_vocab_expansion = _build_diverse_augmentation_pool(
+        paraphrase_pool, training_data_vocab_per_intent
+    )
+    random_expansion = _build_random_augmentation_pool(paraphrase_pool)
+
+    # Build augmentation training set
+    augmented_data_diverse, augmented_data_random = _build_augmentation_training_sets(
+        nlu_training_data,
+        training_data_pool,
+        random_expansion,
+        max_vocab_expansion,
+        pooled_intents,
+        avg_size,
+    )
+
+    # Store training data files
+    output_directory_diverse = os.path.join(args.out, "augmentation_diverse")
+    if not os.path.exists(output_directory_diverse):
+        os.makedirs(output_directory_diverse)
```

Comment:
I think we have a utils function that does this already (create paths).

Reply:
There are (at least) two functions inside rasa that do this, one is `rasa.shared.utils.io.create_directory(...)` the other is `rasa.utils.io.create_path(...)`. The former would already be imported in `rasa/cli/data.py`, though I'm wondering whether there is a preference to use the one over the other?

Reply:
I believe the difference is that `create_directory` takes a path to a directory and creates it along with all parents. On the other hand, `create_path` takes a path to a file and creates all necessary parent directories.